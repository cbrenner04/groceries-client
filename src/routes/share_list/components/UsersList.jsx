import React from 'react';
import PropTypes from 'prop-types';
import { Badge, Button, ButtonGroup, Col, ListGroup, Row } from 'react-bootstrap';

import { Trash } from '../../../components/ActionButtons';
import TitlePopover from '../../../components/TitlePopover';
import { usersLists } from '../../../types';

const UsersList = (props) => (
  <>
    <TitlePopover
      title={props.status}
      message="Click the arrows to upgrade or downgrade the permissions between read and write"
    />
    <ListGroup className="mb-4">
      {props.users.map(({ user, users_list: { id, permissions } }) => {
        if (user.id === props.userId) {
          return '';
        }
        if (props.userIsOwner) {
          return (
            <ListGroup.Item key={id} data-test-id={`${props.status}-user-${user.id}`} style={{ display: 'block' }}>
              <Row>
                <Col md="6" className="pt-1">
                  {user.email}
                </Col>
                <Col md="4" className="pt-1">
                  <Badge data-test-id={`perm-${permissions}`} variant={permissions === 'write' ? 'success' : 'primary'}>
                    {permissions}
                  </Badge>
                </Col>
                <Col md="2">
                  <ButtonGroup className="float-right">
                    <>
                      <Button
                        variant="link"
                        className="p-0 mr-4"
                        onClick={() => props.togglePermission(id, permissions, props.status)}
                        data-test-id="toggle-permissions"
                      >
                        <i
                          className={`fas fa-angle-double-${
                            permissions === 'write' ? 'down' : 'up'
                          } fa-2x text-warning`}
                        />
                      </Button>
                      <Trash testID="remove-share" handleClick={() => props.removeShare(id)} />
                    </>
                  </ButtonGroup>
                </Col>
              </Row>
            </ListGroup.Item>
          );
        }
        return (
          <div key={id} data-test-id={`${props.status}-user-${user.id}`}>
            <ListGroup.Item>{user.email}</ListGroup.Item>
          </div>
        );
      })}
    </ListGroup>
  </>
);

UsersList.propTypes = {
  togglePermission: PropTypes.func.isRequired,
  removeShare: PropTypes.func.isRequired,
  userIsOwner: PropTypes.bool.isRequired,
  userId: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  users: PropTypes.arrayOf(usersLists).isRequired,
};

export default UsersList;
