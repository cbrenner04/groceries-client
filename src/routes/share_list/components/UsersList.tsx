import React from 'react';
import { Badge, Button, ButtonGroup, Col, ListGroup, Row } from 'react-bootstrap';

import { Trash } from 'components/ActionButtons';
import TitlePopover from 'components/TitlePopover';
import type { IUsersList } from 'typings';

export interface IUsersListProps {
  togglePermission: (id: string, permissions: string, status: string) => void;
  removeShare: (id: string) => void;
  userIsOwner: boolean;
  userId: string;
  status: string;
  users: IUsersList[];
}

const UsersList: React.FC<IUsersListProps> = (props): React.JSX.Element => (
  <React.Fragment>
    <TitlePopover
      title={props.status}
      message="Click the arrows to upgrade or downgrade the permissions between read and write"
    />
    <ListGroup className="mb-4">
      {props.users.map((user) => {
        if (user.user.id === props.userId) {
          return '';
        }
        const { permissions } = user.users_list;
        if (props.userIsOwner) {
          return (
            <ListGroup.Item
              key={user.users_list.id}
              data-test-id={`${props.status}-user-${user.user.id}`}
              className="users-list-list-group-item"
            >
              <Row>
                <Col md="6" className="pt-1">
                  {user.user.email}
                </Col>
                <Col md="4" className="pt-1">
                  <Badge data-test-id={`perm-${permissions}`} bg={permissions === 'write' ? 'success' : 'primary'}>
                    {permissions}
                  </Badge>
                </Col>
                <Col md="2">
                  <ButtonGroup className="float-end">
                    <Button
                      variant="link"
                      className="p-0 me-4"
                      onClick={(): void => props.togglePermission(user.users_list.id, permissions, props.status)}
                      data-test-id="toggle-permissions"
                    >
                      <i
                        className={`fas fa-angle-double-${permissions === 'write' ? 'down' : 'up'} fa-2x text-warning`}
                      />
                    </Button>
                    <Trash testID="remove-share" handleClick={(): void => props.removeShare(user.users_list.id)} />
                  </ButtonGroup>
                </Col>
              </Row>
            </ListGroup.Item>
          );
        }
        return (
          <div key={user.users_list.id} data-test-id={`${props.status}-user-${user.user.id}`}>
            <ListGroup.Item>{user.user.email}</ListGroup.Item>
          </div>
        );
      })}
    </ListGroup>
  </React.Fragment>
);

export default UsersList;
