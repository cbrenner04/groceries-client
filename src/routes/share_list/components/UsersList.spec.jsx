import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import UsersList from './UsersList';

describe('UsersList', () => {
  let props;

  beforeEach(() => {
    props = {
      togglePermission: jest.fn(),
      removeShare: jest.fn(),
      userIsOwner: true,
      userId: 'id1',
      status: 'accepted',
      users: [
        {
          user: {
            id: 'id1',
            email: 'foo@example.com',
          },
          users_list: {
            id: 'id1',
            permissions: 'write',
          },
        },
        {
          user: {
            id: 'id2',
            email: 'bar@example.com',
          },
          users_list: {
            id: 'id2',
            permissions: 'write',
          },
        },
        {
          user: {
            id: 'id3',
            email: 'baz@example.com',
          },
          users_list: {
            id: 'id4',
            permissions: 'read',
          },
        },
      ],
    };
  });

  it('renders read and write badges when user is owner', () => {
    props.userIsOwner = true;
    const { container, getByTestId, queryByTestId } = render(<UsersList {...props} />);

    expect(container).toMatchSnapshot();
    expect(queryByTestId('accepted-user-id1')).toBeNull();
    expect(getByTestId('accepted-user-id2').firstChild.children[1].firstChild).toHaveAttribute(
      'data-test-id',
      'perm-write',
    );
    expect(getByTestId('accepted-user-id2').firstChild.children[1].firstChild).toHaveClass('badge-success');
    expect(getByTestId('accepted-user-id3').firstChild.children[1].firstChild).toHaveAttribute(
      'data-test-id',
      'perm-read',
    );
    expect(getByTestId('accepted-user-id3').firstChild.children[1].firstChild).toHaveClass('badge-primary');
  });

  it('does not render read and write badges when user is not owner', () => {
    props.userIsOwner = false;
    const { container, getByTestId, queryByText } = render(<UsersList {...props} />);

    expect(container).toMatchSnapshot();
    expect(queryByText('foo@example.com')).toBeNull();
    expect(getByTestId('accepted-user-id2')).toHaveTextContent('bar@example.com');
    expect(getByTestId('accepted-user-id2')).not.toHaveTextContent('write');
    expect(getByTestId('accepted-user-id3')).toHaveTextContent('baz@example.com');
    expect(getByTestId('accepted-user-id3')).not.toHaveTextContent('read');
  });

  it('toggles user permissions', async () => {
    props.userIsOwner = true;
    const { getAllByRole } = render(<UsersList {...props} />);

    fireEvent.click(getAllByRole('button')[0]);

    // the first user in the list is the signed in user and is therefore not displayed
    const firstDisplayedUser = props.users[1];
    expect(props.togglePermission).toHaveBeenCalledWith(
      firstDisplayedUser.users_list.id,
      firstDisplayedUser.users_list.permissions,
      props.status,
    );
  });

  it('toggles user permissions', async () => {
    props.userIsOwner = true;
    const { getAllByRole } = render(<UsersList {...props} />);

    fireEvent.click(getAllByRole('button')[1]);

    // the first user in the list is the signed in user and is therefore not displayed
    const firstDisplayedUser = props.users[1];
    expect(props.removeShare).toHaveBeenCalledWith(firstDisplayedUser.users_list.id);
  });
});
