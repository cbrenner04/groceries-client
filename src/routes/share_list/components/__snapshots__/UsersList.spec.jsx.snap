// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList does not render read and write badges when user is not owner 1`] = `
<div>
  <div
    class="row"
    style="margin: 0px;"
  >
    <h2
      class="text-capitalize"
    >
      accepted
    </h2>
    <i
      class="far fa-question-circle float-right text-secondary"
      data-test-id="accepted-popover"
    />
  </div>
  <div
    class="mb-4 list-group"
  >
    
    <div
      data-test-id="accepted-user-id2"
    >
      <div
        class="list-group-item"
      >
        bar@example.com
      </div>
    </div>
    <div
      data-test-id="accepted-user-id3"
    >
      <div
        class="list-group-item"
      >
        baz@example.com
      </div>
    </div>
  </div>
</div>
`;

exports[`UsersList renders read and write badges when user is owner 1`] = `
<div>
  <div
    class="row"
    style="margin: 0px;"
  >
    <h2
      class="text-capitalize"
    >
      accepted
    </h2>
    <i
      class="far fa-question-circle float-right text-secondary"
      data-test-id="accepted-popover"
    />
  </div>
  <div
    class="mb-4 list-group"
  >
    
    <div
      class="list-group-item"
      data-test-id="accepted-user-id2"
      style="display: block;"
    >
      <div
        class="row"
      >
        <div
          class="pt-1 col-md-6"
        >
          bar@example.com
        </div>
        <div
          class="pt-1 col-md-4"
        >
          <span
            class="badge badge-success"
            data-test-id="perm-write"
          >
            write
          </span>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="float-right btn-group"
            role="group"
          >
            <button
              class="p-0 mr-4 btn btn-link"
              data-test-id="toggle-permissions"
              type="button"
            >
              <i
                class="fas fa-angle-double-down fa-2x text-warning"
              />
            </button>
            <button
              class="p-0 btn btn-link"
              data-test-id="remove-share"
              type="button"
            >
              <i
                class="fa fa-trash fa-2x text-danger"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
      data-test-id="accepted-user-id3"
      style="display: block;"
    >
      <div
        class="row"
      >
        <div
          class="pt-1 col-md-6"
        >
          baz@example.com
        </div>
        <div
          class="pt-1 col-md-4"
        >
          <span
            class="badge badge-primary"
            data-test-id="perm-read"
          >
            read
          </span>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="float-right btn-group"
            role="group"
          >
            <button
              class="p-0 mr-4 btn btn-link"
              data-test-id="toggle-permissions"
              type="button"
            >
              <i
                class="fas fa-angle-double-up fa-2x text-warning"
              />
            </button>
            <button
              class="p-0 btn btn-link"
              data-test-id="remove-share"
              type="button"
            >
              <i
                class="fa fa-trash fa-2x text-danger"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
